### PR TITLE
Yizha/application insight

### PR DESCRIPTION
**Pull request by yingzhao-1**

### CODE REVIEW

File: src/actions/authActions.js [hunk 1]
```diff
@@ -3,14 +3,22 @@ export const USER_LOGGED_IN = 'USER_LOGGED_IN'
 export const USER_LOGGED_OUT = 'USER_LOGGED_OUT'
 export const USER_LOGIN_ERROR = 'USER_LOGIN_ERROR'
 
+var registerAuthenticatedId = (signInId) => {
```
Comment (pdimitratos) [inline src/actions/authActions.js]: If this is used only within the userLoggedIn function, it might be best to define it within that function. 

Also, prefer const over var for variables when you don't intend for them to be changed after initially being defined (which is generally the case for functions), since that reduces the potential complexity of the code.

File: src/appInsights.js [hunk 2]
```diff
@@ -0,0 +1,9 @@
+import config from 'config'
+
+var appInsights=window.appInsights||function(a){
+    function b(a){c[a]=function(){var b=arguments;c.queue.push(function(){c[a].apply(c,b)})}}var c={config:a},d=document,e=window;setTimeout(function(){var b=d.createElement("script");b.src=a.url||"https://az416426.vo.msecnd.net/scripts/a/ai.0.js",d.getElementsByTagName("script")[0].parentNode.appendChild(b)});try{c.cookie=d.cookie}catch(a){}c.queue=[];for(var f=["Event","Exception","Metric","PageView","Trace","Dependency"];f.length;)b("track"+f.pop());if(b("setAuthenticatedUserContext"),b("clearAuthenticatedUserContext"),b("startTrackEvent"),b("stopTrackEvent"),b("startTrackPage"),b("stopTrackPage"),b("flush"),!a.disableExceptionTracking){f="onerror",b("_"+f);var g=e[f];e[f]=function(a,b,d,e,h){var i=g&&g(a,b,d,e,h);return!0!==i&&c["_"+f](a,b,d,e,h),i}}return c
```
Comment (pdimitratos) [inline src/appInsights.js]: If you wrote this, please provide comments explaining what's going on. If Application insights documentation provided this, please add a comment with a link to that documentation page. This is complex enough that I wouldn't even know where to begin trying to debug it if something causes it to break.

File: src/services/adalService.js [hunk 3]
```diff
@@ -26,12 +26,15 @@ export const getAuthContext = (dispatch) => {
 export const login = (dispatch) => {
   if (typeof window !== 'undefined' && !!window) {
     getAuthContext(dispatch).login()
+
+    appInsights.trackEvent("SIA Login")
```
Comment (pdimitratos) [inline src/services/adalService.js]: These should be tracked in authNService rather than adalService, so we don't have to add tracking in each of adalService, msalService, and any other authentication service options we want to add. The statements here can be moved to the login and logOut methods in authNService before the switch statements to record events for all authentication types.

File: src/actions/authActions.js [hunk 4]
```diff
@@ -1,16 +1,26 @@
...
 export const loginInProgress = () => ({
   type: LOGIN_IN_PROGRESS
 })
 
-export const userLoggedIn = (user) => ({
-  type: USER_LOGGED_IN,
-  user
-})
+export const userLoggedIn = (user) =>{
+  if(config.useAppInsight){
```
Comment (pdimitratos) [inline src/actions/authActions.js]: If we're going to make the use of application insights configurable (and we should), it might be best to check the condition only once in appinsightsService, that way we don't forget to put a guard clause around one of the calls to appinsights, and that way we don't need to import config everywhere we're importing appinsights. It might even be best to wrap the appinsights service in a service similar to AuthNService wrapping ADAL to allow the use of different telemetry platforms. That may be beyond the scope of this pull request, though.

File: src/index.js [hunk 5]
```diff
@@ -5,6 +5,7 @@ import { AppContainer as HotContainer } from 'react-hot-loader'
 
 import { store, persistor } from 'configureStore'
 import MainComponent from 'components/MainComponent'
+import 'appInsights'
```
Comment (pdimitratos) [inline src/index.js]: This should probably change to services/appInsightsService

File: src/appInsights.js [hunk 6]
```diff
@@ -0,0 +1,17 @@
+import config from 'config'
+/*
+To collect end-user usage analytics about your application, 
+insert the following script into each page you want to track.
+Place this code immediately before the closing </head> tag,
+and before any other scripts. Your first data will appear 
+automatically in just a few seconds.
+Please find more info here: https://docs.microsoft.com/en-us/azure/application-insights/app-insights-javascript
```
Comment (pdimitratos) [inline src/appInsights.js]: http://apmtips.com/blog/2015/03/18/javascript-snippet-explained/ has more detailed info on this particular snippet.

This file can likely be merged into appInsightsService; if it can't, then appInsights can be exported from here instead of appInsightsService.

File: src/services/appInsightsService.js [hunk 7]
```diff
@@ -0,0 +1,2 @@
+// eslint-disable-next-line no-undef
+export default appInsights
```
Comment (pdimitratos) [inline src/services/appInsightsService.js]: The appInsights instance is likely accessible in this scope as window.appInsights. That said, it's probably best to merge this into src/appInsights.js or to merge that file into this.

File: src/components/Timeline/Playbook/Play.js [hunk 8]
```diff
@@ -2,9 +2,16 @@ import React from 'react'
 import { connect } from 'react-redux'
 import FlatButtonStyled from 'components/elements/FlatButtonStyled'
 import { fillTemplate, publishEvent } from 'services/playbookService'
+import config from 'config'
```
Comment (pdimitratos) [inline src/components/Timeline/Playbook/Play.js]: Shouldn't need config here anymore

File: src/components/Timeline/Playbook/Play.js [hunk 9]
```diff
@@ -2,9 +2,16 @@ import React from 'react'
...
+import appInsights from 'src/appInsights'
+
+var trackLinkClick = (name) => {
+  // eslint-disable-next-line no-undef
```
Comment (pdimitratos) [inline src/components/Timeline/Playbook/Play.js]: Shouldn't need the eslint directive now that appInsights is imported rather than assumed to be in the global scope

File: cfg/base.js [hunk 10]
```diff
@@ -35,11 +36,11 @@ const config = {
     extensions: ['.js'],
     alias: {
       actions: `${srcPath}/actions/`,
-      components: `${srcPath}/components/`,
```
Comment (macrobug) [inline cfg/base.js]: why this change?

File: cfg/base.js [hunk 11]
```diff
@@ -35,11 +36,11 @@ const config = {
...
       sources: `${srcPath}/sources/`,
       stores: `${srcPath}/stores/`,
       styles: `${srcPath}/styles/`,
       config: `${srcPath}/config/`,
+      src: `${srcPath}/`,
```
Comment (macrobug) [inline cfg/base.js]: why this change?

File: src/actions/authActions.js [hunk 12]
```diff
@@ -1,16 +1,24 @@
+import config from 'config'
+//import appInsights from 'src/appInsights'
```
Comment (macrobug) [inline src/actions/authActions.js]: remove commented out code

File: src/actions/authActions.js [hunk 13]
```diff
@@ -1,16 +1,24 @@
...
 export const loginInProgress = () => ({
   type: LOGIN_IN_PROGRESS
 })
 
-export const userLoggedIn = (user) => ({
-  type: USER_LOGGED_IN,
-  user
-})
+export const userLoggedIn = (user) =>{
+  appInsights.setAuthenticatedUserContext(user.userName.replace(/[,;=| ]+/g, "_"), null, true);
```
Comment (macrobug) [inline src/actions/authActions.js]: is this the right place for this code? consider moving to the reducer. consult with Philip what is the right approach.

File: src/components/Timeline/Playbook/Play.js [hunk 14]
```diff
@@ -2,9 +2,16 @@ import React from 'react'
 import { connect } from 'react-redux'
 import FlatButtonStyled from 'components/elements/FlatButtonStyled'
 import { fillTemplate, publishEvent } from 'services/playbookService'
+import config from 'config'
+//import appInsights from 'src/appInsights'
```
Comment (macrobug) [inline src/components/Timeline/Playbook/Play.js]: remove commented out code

File: src/components/Timeline/Playbook/Play.js [hunk 15]
```diff
@@ -2,9 +2,16 @@ import React from 'react'
...
+import appInsights from '../../../appInsights'
```
Comment (macrobug) [inline src/components/Timeline/Playbook/Play.js]: remove ../../../ relative path. this should work: 'services/appInsights'

File: src/components/Timeline/Playbook/Play.js [hunk 16]
```diff
@@ -2,9 +2,16 @@ import React from 'react'
...
+
+var trackLinkClick = (name) => {
```
Comment (macrobug) [inline src/components/Timeline/Playbook/Play.js]: since trackLinkClick never changes, use `const`

File: src/index.html [hunk 17]
```diff
@@ -1,5 +1,6 @@
 <!doctype html>
 <html>
+  <script type="text/javascript" src="/assets/appInsights.js"></script>
```
Comment (macrobug) [inline src/index.html]: According to the appinsights docs you pasted earlier, this should go before `</head> `not before `<head>`

File: cfg/defaultConstants.js [hunk 18]
```diff
@@ -39,5 +39,9 @@ module.exports = {
...
       ticketUriSuffix: '&edit=false'
     }
-  }
-}
+  },
+
+  instrumentationKey: 'APP INSIGHT GOES HERE',
+  
+  useAppInsights: true
+}
```
Comment (pdimitratos) [inline cfg/defaultConstants.js]: Don't forget to end files with a line break for git compatibility.

File: src/index.html (hunk 17 above)
Comment (yingzhao-1) [reply src/index.html]: Application insights works correctly by injecting the codes here. So it should be fine leaving it this way.

File: src/components/Timeline/Playbook/Play.js (hunk 15 above)
Comment (yingzhao-1) [reply src/components/Timeline/Playbook/Play.js]: In which file was 'services' defined? I cannot find it.

Comment (yingzhao-1) [reply src/components/Timeline/Playbook/Play.js]: And appInsights.js is not under services.
